nibabel/streamlines/array_sequence.py

```diff
@@ -0,0 +1,260 @@
+            e = np.asarray(e)
+            if i == 0:
+                new_shape = (ArraySequence.BUFFER_SIZE,) + e.shape[1:]
+                self._data = np.empty(new_shape, dtype=e.dtype)
+
+            end = offset + len(e)
+            if end >= len(self._data):
+                # Resize needed, adding `len(e)` items plus some buffer.
+                nb_points = len(self._data)
+                nb_points += len(e) + ArraySequence.BUFFER_SIZE
+                self._data.resize((nb_points,) + self.common_shape)
+
+            offsets.append(offset)
+            lengths.append(len(e))
+            self._data[offset:offset + len(e)] = e
+            offset += len(e)
+
+        self._offsets = np.asarray(offsets)
+        self._lengths = np.asarray(lengths)
+
+        # Clear unused memory.
+        self._data.resize((offset,) + self.common_shape)
+
+    @property
+    def is_array_sequence(self):
+        return True
+
+    @property
+    def common_shape(self):
+        """ Matching shape of the elements in this array sequence. """
+        return self._data.shape[1:]
+
+    def append(self, element):
+        """ Appends `element` to this array sequence.
+
+        Parameters
+        ----------
+        element : ndarray
+            Element to append. The shape must match already inserted elements
+            shape except for the first dimension.
+
+        Notes
+        -----
+        If you need to add multiple elements you should consider
+        `ArraySequence.extend`.
+        """
+        element = np.asarray(element)
+
+        if self.common_shape != () and element.shape[1:] != self.common_shape:
+            msg = "All dimensions, except the first one, must match exactly"
+            raise ValueError(msg)
+
+        next_offset = self._data.shape[0]
+        size = (self._data.shape[0] + element.shape[0],) + element.shape[1:]
+        self._data.resize(size)
+        self._data[next_offset:] = element
+        self._offsets = np.r_[self._offsets, next_offset]
+        self._lengths = np.r_[self._lengths, element.shape[0]]
+
+    def extend(self, elements):
+        """ Appends all `elements` to this array sequence.
+
+        Parameters
+        ----------
+        elements : iterable of ndarrays or :class:`ArraySequence` object
+            If iterable of ndarrays, each ndarray will be concatenated along
+            the first dimension then appended to the data of this
+            ArraySequence.
+            If :class:`ArraySequence` object, its data are simply appended to
+            the data of this ArraySequence.
+
+        Notes
+        -----
+            The shape of the elements to be added must match the one of the
+            data of this :class:`ArraySequence` except for the first dimension.
+        """
+        if not is_array_sequence(elements):
+            self.extend(ArraySequence(elements))
+            return
+
+        if len(elements) == 0:
+            return
+
+        if (self.common_shape != () and
```

Comment:
What do you mean exactly? `extend` is using code from the `__init__` method by calling it explicitly when possible. Would you prefer I move the `__init__` code into to extend, and in `__init__` call the extend method?